src/puppetlabs/puppetdb/amq_migration.clj

```diff
@@ -0,0 +1,369 @@
+
+(defn consume-everything
+  "Attempts to read all available messages from ActiveMQ, giving up
+  after 5 seconds if one is not available"
+  [^MessageConsumer consumer process-message id]
+  (try
+    (loop []
+      (when-let [msg (.receive consumer 5000)]
+        (process-message (swap! id inc) msg)
+        (recur)))
+    (catch javax.jms.IllegalStateException e
+      (log/info (i18n/trs "Received IllegalStateException, shutting down")))
+    (catch Exception e
+      (log/error e))))
+
+(defn create-mq-receiver
+  [^ConnectionFactory conn-pool endpoint process-message id]
+  (with-open [connection (create-connection conn-pool)
+              session (create-session connection)
+              consumer (.createConsumer session (create-queue session endpoint))]
+    (log/info (i18n/trs "Processing messages from the Queue: {0}" endpoint))
+    (consume-everything consumer process-message id)))
+
+(defn create-scheduler-receiver
+  "Drains the scheduler queue (where PuppetDB commands are retried)
+  and enqueues them into the normal queue"
+  [^ConnectionFactory conn-pool endpoint process-message dlo id]
+  (with-open [connection (create-connection conn-pool)
+              session (create-session connection)]
+    (let [request-browse (.createTopic session ScheduledMessage/AMQ_SCHEDULER_MANAGEMENT_DESTINATION)
+          browse-dest (.createTemporaryQueue session)]
+      (with-open [producer (.createProducer session request-browse)
+                  consumer (.createConsumer session browse-dest)]
+        (let [request (doto (.createMessage session)
+                        (.setStringProperty ScheduledMessage/AMQ_SCHEDULER_ACTION
+                                            ScheduledMessage/AMQ_SCHEDULER_ACTION_BROWSE)
+                        (.setJMSReplyTo browse-dest))
+              process-message-and-remove (fn [^Message msg]
+                                           (try
+                                             (process-message msg)
+                                             (let [msg-id (.getStringProperty msg
+                                                                              ScheduledMessage/AMQ_SCHEDULED_ID)
+                                                   remove (doto (.createMessage session)
+                                                            (.setStringProperty ScheduledMessage/AMQ_SCHEDULER_ACTION
+                                                                                ScheduledMessage/AMQ_SCHEDULER_ACTION_REMOVE)
+                                                            (.setStringProperty ScheduledMessage/AMQ_SCHEDULED_ID msg-id))]
+                                               (.send producer remove))
+                                             (catch Exception e
+                                               (discard-amq-message (convert-message-body msg) (swap! id inc) {} e dlo))))]
+          (.send producer request)
+          (Thread/sleep 2000)
```

Comment:
I think the comment got wiped, I asked @ajroetker and didn't hear back, trying a second ping.